cache/lru/lru.go

```diff
@@ -166,3 +166,52 @@ func (c *Cache[K, V]) Len() int {
 
 	return c.ll.Len()
 }
+
+// Delete removes an item from the cache.
+func (c *Cache[K, V]) Delete(key K) {
+	c.LoadAndDelete(key)
+}
+
+// LoadAndDelete queries an item and deletes it from the cache using the
+// specified key.
+func (c *Cache[K, V]) LoadAndDelete(key K) (V, bool) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	var defaultVal V
+
+	// Noop if the element doesn't exist.
+	el, ok := c.cache[key]
+	if !ok {
+		return defaultVal, false
+	}
+
+	// Before we delete the element, we need to get its size.
+	vs, err := el.Value.value.Size()
+	if err != nil {
+		return defaultVal, false
+	}
+
+	// Remove the element from the list.
+	c.ll.Remove(el)
+
+	// Delete the element from the cache and update the cache's size.
+	delete(c.cache, key)
+	c.size -= vs
+
+	return el.Value.value, true
+}
+
+// Range iterates the cache.
+//
+// NOTE: the `visitor` is wrapped inside the Cache's mutex.
+func (c *Cache[K, V]) Range(visitor func(K, V) bool) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	for k, v := range c.cache {
+		if !visitor(k, v.Value.value) {
```

Comment:
Hmm, one thing that pops up here is that since Go doesn't have reentrant locks, then if `visitor` tries to call any of these methods, it'll deadlock.

Reply:
yeah I also think this design might be dangerous. Some thoughts,
- maybe we could copy the `cache` inside a lock above and then iterate? Meanwhile the `cache` could be updated tho by another goroutine tho.
- Remove the lock in `Range`, this means the caller needs to remember using the locks.
- Replace `cache` with a sync map, lower performance on write, but no more worries on locks.

Reply:
> Replace cache with a sync map, lower performance on write, but no more worries on locks.

I think this could work, w/o forcing us to reimplement the wheel too much.

Reply:
Cool, used `SyncMap` with code copied from [here](https://github.com/lightningnetwork/lnd/blob/master/lnutils/sync_map.go). We may want to create a repo to host these commonly used utilities someday.

The `mutex` used by the `cache` is still needed tho, to update its internal list and size.

Reply:
> We may want to create a repo to host these commonly used utilities someday.

Alternatively, assuming we keep `lnutils` w/ a slim set of deps, we can make that into a sub-module so other repos can call into it w/o having to worry about weird import cycle.